Update reference docs to detail universal deletion policy (#17445)
GoogleCloudPlatform/magic-modules: docs/content/reference/universal-deletion-policy.md

---

> title: "Universal Deletion Policy"

Reviewer: I'd recommend removing references to "Universal" here & elsewhere (including the filename) - although it's the name of the project, it's probably not important to contributors.

Additionally - this page probably shouldn't be in `reference`, since it's not a reference - it should either go in `develop` or `best-practices`. Personally I'd lean towards replacing `best-practices/deletion-behaviors` with this content. You don't need to keep that page name, but make sure to add an alias so that references to the old location will redirect.

You'll also want to update the [MMv1 resource reference](https://googlecloudplatform.github.io/magic-modules/reference/resource/) to include the new YAML fields in a reference format (similar to the way other YAML fields are documented on that page.)

weight: 20
---

> Almost all resources in the provider now support a universal `deletion_policy` field, superceding previous guidance on implementing `deletion_protection` or unique `deletion_policy` fields from [Deletion behaviors]({{< ref "/best-practices/deletion-behaviors" >}}).

Reviewer: here and elsewhere: I'd recommend just capturing the current state; we don't need to refer to the previous guidance - or if we do, it should be deprioritized (for example, at the end this page, you could have a "Legacy deletion behaviors" section that summarizes the important points that people need to know for maintaining older fields on resources.)

This page documents how to work with the universal deletion policy in either generated or handwritten resources.

{{% tabs "schema" %}}
{{< tab "MMv1" >}}
By default, existing and new resources should support `deletion_policy` without any complication or changes needed by a contributor. A handful of options are available to customize the implementation as necessary.

##Updating the default value of `deletion_policy` for a resource
All resources will use a value of "DELETE" by default if unspecified, but this default can be changed to another value by including the following:
```yaml
deletion_policy_default: "PREVENT" #this can be any string
```
This is how resources should be modified for implementing default deletion protection guards going forward.

##Supporting a custom value for deletion_policy
If implementing support for additional values such as "FORCE", the following steps can be taken.

> A `pre_delete` constant should be added to the `custom_code` block in the resource YAML file, that performs the logic with the corresponding value. Example implemention here: [#1](https://github.com/GoogleCloudPlatform/magic-modules/FEATURE-BRANCH-universal-deletion-policy/mmv1/products/datastream/PrivateConnection.yaml#L44), [#2](https://github.com/GoogleCloudPlatform/magic-modules/blob/FEATURE-BRANCH-universal-deletion-policy/mmv1/templates/terraform/pre_delete/private_connection.go.tmpl)

Reviewer: Generally we try to avoid "should": https://developers.google.com/style/prescriptive-documentation#word-choice

I think this might be more clear in active voice as well: https://developers.google.com/style/voice

```suggestion
Add a `pre_delete` constant to the `custom_code` block in the resource YAML file that performs the logic with the corresponding value. Example implemention here: [#1](https://github.com/GoogleCloudPlatform/magic-modules/FEATURE-BRANCH-universal-deletion-policy/mmv1/products/datastream/PrivateConnection.yaml#L44), [#2](https://github.com/GoogleCloudPlatform/magic-modules/blob/FEATURE-BRANCH-universal-deletion-policy/mmv1/templates/terraform/pre_delete/private_connection.go.tmpl)
```

Also relevant: https://developers.google.com/style/procedures

This should also be changed in the preceding & following sentences, and maybe elsewhere.
